core/domain/exp_domain_test.py

```diff
@@ -429,6 +429,493 @@ def test_validate_exp_id(self):
+            self.exploration.init_state_name: self.new_state
+        }
+        self.set_interaction_for_state(
+            self.exploration.states[self.exploration.init_state_name],
+            'TextInput')
+        self.init_state = (
+            self.exploration.states[self.exploration.init_state_name])
+        self.end_state = state_domain.State.create_default_state('End')
+        self.set_interaction_for_state(self.end_state, 'EndExploration')
+
+        self.end_state.update_interaction_default_outcome(None)
+
+    def test_init_state_with_card_is_checkpoint_false_is_invalid(self):
+        self.init_state.update_card_is_checkpoint(False)
+        self._assert_validation_error(
+            self.exploration, 'Expected card_is_checkpoint of first state to '
+            'be True but found it to be False')
+        self.init_state.update_card_is_checkpoint(True)
+        self.exploration.validate()
+
+    def test_end_state_with_card_is_checkpoint_true_is_invalid(self):
+        default_outcome = self.init_state.interaction.default_outcome
+        default_outcome.dest = self.exploration.init_state_name
+        self.init_state.update_interaction_default_outcome(default_outcome)
+
+        self.exploration.states = {
+            self.exploration.init_state_name: self.new_state,
+            'End': self.end_state
+        }
+        self.end_state.update_card_is_checkpoint(True)
+        self._assert_validation_error(
+            self.exploration, 'Expected card_is_checkpoint of terminal state '
+            'to be False but found it to be True')
+        self.end_state.update_card_is_checkpoint(False)
+        self.exploration.validate()
+
+    def test_init_state_checkpoint_with_end_exp_interaction_is_valid(self):
+        self.exploration.init_state_name = 'End'
+        self.exploration.states = {
+            self.exploration.init_state_name: self.end_state
+        }
+        self.end_state.update_card_is_checkpoint(True)
+        self.exploration.validate()
+        self.end_state.update_card_is_checkpoint(False)
+
+    def test_checkpoint_count_with_count_outside_range_is_invalid(self):
+        self.exploration.init_state_name = 'Introduction'
+        self.exploration.states = {
+            self.exploration.init_state_name: self.new_state,
+            'End': self.end_state
+        }
+
+        for i in python_utils.RANGE(8):
+            self.exploration.add_states(['State%s' % i])
+            self.exploration.states['State%s' % i].card_is_checkpoint = True
+        self._assert_validation_error(
+            self.exploration, 'Expected checkpoint count to be between 1 and 8 '
+            'inclusive but found it to be 9'
+        )
+        self.exploration.states = {
+            self.exploration.init_state_name: self.new_state,
+            'End': self.end_state
+        }
+        self.exploration.validate()
+
+    def test_bypassable_state_with_card_is_checkpoint_true_is_invalid(self):
```

Comment:
This is a very long test which is quite hard to parse. I suggest that you add ASCII art or a simple "words and arrows" representation for schematic graphs of the explorations you're constructing here, and add line comments that break the test up into meaningful sections and that explain what each section is doing -- or, better still, have each type of graph in its own separate unit test that explains what specifically is being tested in that test.

Reply:
Added ASCII representation of graphs along with some comments. Is the test now easier to parse?